pkg/users/local.js

```diff
@@ -37,18 +38,67 @@ function AccountsPage() {
     const groups = useFile("/etc/group", { syntax: etc_group_syntax });
     const current_user_info = useLoggedInUser();
 
-    if (!accounts || !groups || !current_user_info)
+    const [details, setDetails] = useState(null);
+    useEffect(() => {
+        getLogins().then(setDetails);
+
+        // Watch `/var/run/utmp` to register when user logs in or out
+        const handle = cockpit.file("/var/run/utmp", { superuser: "try", binary: true });
+        handle.watch(() => {
+            getLogins().then(setDetails);
+        });
+        return handle.close;
+    }, [path, accounts, shadow]);
+
+    if (!accounts || !groups || !current_user_info || !details)
         return null;
 
-    if (path.length === 0)
-        return <AccountsList accounts={accounts} current_user={current_user_info.name} />;
-    else
+    // lastlog uses same sorting as /etc/passwd therefore arrays can be combined based on index
+    const accountsInfo = accounts.map((account, i) => {
+        return Object.assign({}, account, details[i]);
+    });
+
+    if (path.length === 0) {
+        return <AccountsMain accountsInfo={accountsInfo} current_user={current_user_info.name} groups={groups} />;
+    } else
         return (
             <AccountDetails accounts={accounts} groups={groups} shadow={shadow}
                             current_user={current_user_info.name} user={path[0]} />
         );
 }
 
+async function getLogins() {
+    const lastlog = await cockpit.spawn(["/usr/bin/lastlog"], { environ: ["LC_ALL=C"] })
+            .catch(err => console.warn("Unexpected error when getting last login information", err));
+    const w = await cockpit.spawn(["/usr/bin/w", "-sh"], { environ: ["LC_ALL=C"] })
+            .catch(err => console.warn("Unexpected error when getting logged in accounts", err));
+
+    const currentLogins = w.split('\n').slice(0, -1).map(line => {
+        return line.split(/ +/)[0];
+    });
+
+    // drop header and last empty line with slice
+    const promises = lastlog.split('\n').slice(1, -1).map(line => {
+        const splitLine = line.split(/ +/);
+        const name = splitLine[0];
+
+        if (line.indexOf('**Never logged in**') > -1) {
+            return new Promise(resolve => {
+                resolve({ name: name, loggedIn: false, lastLogin: null });
+            });
+        }
+
+        const date_fields = splitLine.slice(-5);
+        return cockpit.spawn(["date", "+%s", "-d", date_fields.join(' ')], { environ: ["LC_ALL=C"], err: "out" })
```

Comment:
Can't this be parsed with `DateTime`? Seems a lot more efficient?. Possibly with `date-fns`'s parse() function?

Reply:
Ooh, looks like I forgot keep the original comment. I'll add it back! The reason `date` is used is explained in [this commit](https://github.com/cockpit-project/cockpit/commit/b81ef3cd24ba2a9427e8e8c4a3997fbc79fb4b5f).

Reply:
Added the comment inline.